File: ingestors/rust/src/trace.rs
use serde::{Deserialize, Serialize};
use std::collections::{HashMap, HashSet};

#[derive(Deserialize, Debug, Serialize)]
pub struct KeyVal {
    pub name: String,
    pub value: String,
}

#[derive(Deserialize, Debug, Serialize)]
pub struct ApiUrl {
    pub host: String,
    pub path: String,
    pub parameters: Vec<KeyVal>,
}

#[derive(Deserialize, Debug, Serialize)]
pub struct ApiRequest {
    pub method: String,
    pub url: ApiUrl,
    pub headers: Vec<KeyVal>,
    pub body: Option<String>,
}

#[derive(Deserialize, Debug, Serialize)]
pub struct ApiResponse {
    pub status: u16,
    pub headers: Vec<KeyVal>,
    pub body: Option<String>,
}

#[derive(Deserialize, Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ApiMeta {
    pub environment: String,
    pub incoming: bool,
    pub source: String,
    pub source_port: u16,
    pub destination: String,
    pub destination_port: u16,
}

#[derive(Deserialize, Debug, Serialize)]
pub struct ApiTrace {
    pub request: ApiRequest,
    pub response: Option<ApiResponse>,
    pub meta: Option<ApiMeta>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ProcessTraceRes {
    pub block: bool,
    pub xss_detected: Option<HashMap<String, String>>,
    pub sqli_detected: Option<HashMap<String, (String, String)>>,
    pub sensitive_data_detected: Option<HashMap<String, HashSet<String>>>,
    pub data_types: Option<HashMap<String, HashSet<String>>>,
    pub validation_errors: Vec<String>,
}

File: ingestors/golang/metlo/trace.go
package metlo

type NV struct {
	Name  string `json:"name"`
	Value string `json:"value"`
}

type TraceUrl struct {
	Host       string `json:"host"`
	Path       string `json:"path"`
	Parameters []NV   `json:"parameters"`
}

type TraceReq struct {
	Url     TraceUrl `json:"url"`
	Headers []NV     `json:"headers"`
	Body    string   `json:"body"`
	Method  string   `json:"method"`
}

type TraceRes struct {
	Status  int    `json:"status"`
	Headers []NV   `json:"headers"`
	Body    string `json:"body"`
}

type TraceMeta struct {
	Environment     string `json:"environment"`
	Incoming        bool   `json:"incoming"`
	Source          string `json:"source"`
	SourcePort      int    `json:"sourcePort"`
	Destination     string `json:"destination"`
	DestinationPort int    `json:"destinationPort"`
	MetloSource     string `json:"metloSource"`
}

type MetloTrace struct {
	Request  TraceReq  `json:"request"`
	Response TraceRes  `json:"response"`
	Meta     TraceMeta `json:"meta"`
}
